src/planning.ts

```diff
@@ -62,284 +63,334 @@ export const defaultPlanOptions: PlanOptions = {
-    const block = this.blocks[blockIdx];
-    return block.instant(t - this.ts[blockIdx], this.ts[blockIdx], this.ss[blockIdx]);
+  public instant (t: number): Instant {
+    const idx = sortedIndex(this.ts, t)
+    const blockIdx = this.ts[idx] === t ? idx : idx - 1
+    const block = this.blocks[blockIdx]
+    return block.instant(t - this.ts[blockIdx], this.ts[blockIdx], this.ss[blockIdx])
   }
 
-  public serialize(): any {
+  public serialize (): any {
     return {
-      t: "XYMotion",
+      t: 'XYMotion',
       blocks: this.blocks.map((b) => b.serialize())
-    };
+    }
   }
 }
 
 export class Plan {
-  public static deserialize(o: any): Plan {
+  public static deserialize (o: any): Plan {
     return new Plan(o.motions.map((m: any) => {
       switch (m.t) {
-        case "XYMotion": return XYMotion.deserialize(m);
-        case "PenMotion": return PenMotion.deserialize(m);
+        case 'XYMotion': return XYMotion.deserialize(m)
+        case 'PenMotion': return PenMotion.deserialize(m)
       }
-    }));
+    }), o.minPenPosition)
   }
-  public motions: Motion[];
-  public constructor(motions: Motion[]) {
-    this.motions = motions;
+
+  private readonly minPenPosition: number
+  public motions: Motion[]
+  public constructor (motions: Motion[], minPenPosition: number) {
+    this.motions = motions
+    this.minPenPosition = minPenPosition
   }
-  public duration(start = 0): number {
-    return this.motions.slice(start).map((m) => m.duration()).reduce((a, b) => a + b, 0);
+
+  public duration (start = 0): number {
+    return this.motions.slice(start).map((m) => m.duration()).reduce((a, b) => a + b, 0)
   }
-  public motion(i: number) { return this.motions[i]; }
 
-  public withPenHeights(penUpHeight: number, penDownHeight: number): Plan {
-    let penMotionIndex = 0;
+  public motion (i: number) { return this.motions[i] }
+
+  public withPenHeights (penUpHeight: number, penDownHeight: number): Plan {
+    let penMotionIndex = 0
     return new Plan(this.motions.map((motion, j) => {
       if (motion instanceof XYMotion) {
-        return motion;
+        return motion
       } else if (motion instanceof PenMotion) {
-        // Uuuugh this is really hacky. We should instead store the
-        // pen-up/pen-down heights in a single place and reference them from
-        // the PenMotions. Then we can change them in just one place.
-        if (j === this.motions.length - 3) {
-          return new PenMotion(penDownHeight, Device.Axidraw.penPctToPos(0), motion.duration());
-        } else if (j === this.motions.length - 1) {
-          return new PenMotion(Device.Axidraw.penPctToPos(0), penUpHeight, motion.duration());
+        // TODO: Remove this hack by storing the pen-up/pen-down heights
+        // in a single place, and reference them from the PenMotions.
+        if (j === this.motions.length - 1) {
+          return new PenMotion(this.minPenPosition, penUpHeight, motion.duration())
         }
```

Comment:
I *think* this is compensating for the additional pen home command we add to our existing plans. Would love to get a more thorough second look here, is this is a logic change.

Reply:
https://github.com/nornagon/saxi/issues/90